web/web.go

```diff
@@ -0,0 +1,773 @@
+	// etc). In particular, note that we can't just blindly redo schema conversion (using an appropriate
+	// version of 'toDDL' with the new type mapping).
+	for t, spSchema := range app.conv.SpSchema {
+		for col, _ := range spSchema.ColDefs {
+			sourceTable := app.conv.ToSource[t].Name
+			sourceCol := app.conv.ToSource[t].Cols[col]
+			srcCol := app.conv.SrcSchema[sourceTable].ColDefs[sourceCol]
+			if _, found := typeMap[srcCol.Type.Name]; found {
+				var ty ddl.Type
+				var issues []internal.SchemaIssue
+				switch app.driver {
+				case "mysql", "mysqldump":
+					ty, issues = toSpannerTypeMySQL(srcCol.Type.Name, typeMap[srcCol.Type.Name], srcCol.Type.Mods)
+				case "pg_dump", "postgres":
+					ty, issues = toSpannerTypePostgres(srcCol.Type.Name, typeMap[srcCol.Type.Name], srcCol.Type.Mods)
+				default:
+					http.Error(w, fmt.Sprintf("Driver : '%s' is not supported", app.driver), http.StatusBadRequest)
+					return
+				}
+				if len(srcCol.Type.ArrayBounds) > 1 {
+					ty = ddl.Type{Name: ddl.String, Len: ddl.MaxLength}
+					issues = append(issues, internal.MultiDimensionalArray)
+				}
+				if srcCol.Ignored.Default {
+					issues = append(issues, internal.DefaultValue)
+				}
+				if srcCol.Ignored.AutoIncrement {
+					issues = append(issues, internal.AutoIncrement)
+				}
+				if len(issues) > 0 {
+					app.conv.Issues[sourceTable][srcCol.Name] = issues
+				}
+				ty.IsArray = len(srcCol.Type.ArrayBounds) == 1
+				colDef := spSchema.ColDefs[col]
+				colDef.T = ty
+				spSchema.ColDefs[col] = colDef
+			}
+		}
+	}
+	updateSessionFile()
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(app.conv)
+}
+
+// Actions to be performed on a column.
+// (1) Removed: true/false
+// (2) Rename: New name or empty string
+// (3) PK: "ADDED", "REMOVED" or ""
+// (4) NotNull: "ADDED", "REMOVED" or ""
+// (5) ToType: New type or empty string
+type updateCol struct {
+	Removed bool   `json:"Removed"`
+	Rename  string `json:"Rename"`
+	PK      string `json:"PK"`
+	NotNull string `json:"NotNull"`
+	ToType  string `json:"ToType"`
+}
+
+type updateTable struct {
+	UpdateCols map[string]updateCol `json:"UpdateCols"`
+}
+
+// updateTableSchema updates the Spanner schema.
+// Following actions can be performed on a specified table:
+// (1) Remove column
+// (2) Rename column
+// (3) Add or Remove Primary Key
+// (4) Add or Remove NotNull constraint
+// (5) Update Spanner type
+func updateTableSchema(w http.ResponseWriter, r *http.Request) {
+	reqBody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Body Read Error : %v", err), http.StatusInternalServerError)
+		return
+	}
+	var t updateTable
+	table := r.FormValue("table")
+	err = json.Unmarshal(reqBody, &t)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Request Body parse error : %v", err), http.StatusBadRequest)
+		return
+	}
+	srcTableName := app.conv.ToSource[table].Name
+	for colName, v := range t.UpdateCols {
```

Comment:
How should we manage errors in the block of code?
At the moment, it looks like we try to do each update step, and any step that can generate an error (such as updateType) will set http.Error. But what happens if we hit multiple errors -- we'd just get the http error for the last one?

I'm not sure what we should do here.
But it might be more intuitive to fail on first error, rather than continuing, and not to do any updates of the session - could we reload the session file?
Also, maybe we should